scripts/test-certificates.sh

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env bash
+set -ueo pipefail
+source $(dirname $0)/utils.sh
+
+TMPDIR=$(mktemp -d -t tmp.XXXXXXXXXX)
+IDENTS_DIR=$TMPDIR/identities
+CERTS_DIR=$TMPDIR/certificates
+# TODO: make port numbers more dynamic
```

Comment:
I'm leaving the "follows testing guide" box unchecked because I think we're supposed to choose ports dynamically but I'm also not totally sure how to do this in bash....

Reply:
Yea, I don't think we have any examples of ports being dynamically chosen in bash but I chose a port range above the storj-sim defaults and it works in jenkins.

Reply:
I've been trying to find a solution but what I tested didn't work.

Reply:
Here's a dumb but portable (macOS and linux) way to get an unused tcp port:

```sh
    port=$(netstat -ant -p tcp 2>/dev/null | awk '$0~/tcp/ {sub(/.*[:.]/,"",$4); a[$4]=1} END {for(i=1024;i<65000;i++) if(a[i]==0) {print i; exit}}')
    [ -n "$port" ] || { echo "No ports available" >&2; exit 1; }
```

Obviously there's a race between choosing the port and using it, but in many systems that is acceptable. To really fix it, the best thing would be to allow passing `:0` to the application so that the tcp stack itself chooses an unused port when opening the socket, and then you make a way for the application to report when it has chosen a port and when it's ready to accept connections. Sometimes this takes the form of an option like `--report-fd=3` meaning the shell will open a pipe on fd 3 and the application is expected to write its status (including the listening port number) to fd 3 when it's ready to go. Or, sometimes, you make the shell do something like `tail -n +1 -F application.logfile | grep -m1 "Listening on port:"` to wait until the application reports its listening port to the log. But both of those are probably more work than they're worth.

Reply:
Use a different localhost to avoid conflicts. These will surely end-up causing flakiness in tests. Go tests use random ports on `127.0.0.1`, other tests use `127.0.0.2` etc.

Reply:
Using different interfaces is a nice and easy shortcut for now, but it's really just a punting of the problem. We will still need, eventually, to keep track of what pieces of the system are using which localhost interfaces to avoid collisions. And if we can do that, presumably we can also keep track of which pieces of the system are using which port numbers..

Reply:
Yeah, I tend to agree. As much as possible tests are using `127.0.0.1:0` and then pick up the actual port being used. This works nicely as long as you can access the listener easily. For processes we do it in some places, e.g. for redis-server. However that's a annoying to work with and requires magic strings to work. Additionally you need to then handle cases that satellite exposes several ports.

Integration tests are pretty isolated and often require a bunch of ports to be set up. e.g. storj-sim integration with default setup ends up using over 35 ports, I think. The port layout of storj-sim is defined to be easier to remember, so they do not map to a sequence of ports. When we run multiple storj-sim-s side-by-side, as we already do, we would need to remap and reconfigure all those 35 ports.

Basically, it's easier to separate interfaces. It's possible, but also annoying. The conflicts also often end up as flakiness.

Reply:
I find Docker to be hugely overused in the industry right now, but for better setup and teardown of isolated test networks that can't step on each other and can't leave surprising background processes lying around, it's really hard to beat containers!